core/src/main/scala/org/apache/spark/sql/delta/util/DeltaFileOperations.scala

```diff
@@ -333,13 +333,20 @@ object DeltaFileOperations extends DeltaLogging {
    * because the file didn't exist, then we still return `true`. Retries on S3 rate limits up to 3
    * times.
    */
-  def tryDeleteNonRecursive(fs: FileSystem, path: Path, tries: Int = 3): Boolean = {
+  def tryDeleteNonRecursive(
+      fs: FileSystem,
+      path: Path,
+      tries: Int = 3,
+      firstCall: Boolean = true): Boolean = {
     try fs.delete(path, false) catch {
       case _: FileNotFoundException => true
+      case ioException: IOException if firstCall =>
```

Comment:
on second thought, why do we need this `if firstCall` check?

Wouldn't this achieve the same result? We don't recurse any more ...

And if we did recurse ... and on the 3rd try ... we DO see an IOException ... then we know that we have not ever received a previous IOException.

```scala
case _: IOException =>
  logError(s"Vacuum failed to delete ${path.toString}", ioException)
  false
```

Reply:
@amirmor1